client/mysql.cc

```diff
@@ -5315,6 +5365,512 @@ static int com_prompt(String *buffer __attribute__((unused)),
+        end= pos;
+        pos ++;
+        break;
+      }
+      while (!my_isspace(charset_info, *pos)) pos ++;
+      end= pos;
+      break;
+    default:
+      /* Its an invalid entry. Lets move until we find a space. */
+      while (!my_isspace(charset_info, *pos)) pos ++;
+      end= pos;
+    }
+  }
+  else
+  {
+    /* We have reached the end. */
+    if (!quoted) *is_valid= true;
+    end= pos;
+  }
+
+  len= end - beg;
+  value= (char *) my_malloc(len + 1, MYF(MY_WME));
+  memcpy(value, beg, len);
+  value[len]= 0;
+  *out= value;
+  return pos;
+}
+
+static char *handle_next_alias(char *line, bool *error)
+{
+  ALIAS *alias;
+  char *name, *value, *pos;
+  bool is_valid;
+  size_t name_len;
+  uchar *record;
+
+  /* Parse the alias name. */
+  pos= parse_alias_name(line, &name, &is_valid);
+
+  if (!is_valid)
+  {
+    tee_fprintf(stdout, "alias: '%s': invalid alias name\n", name);
+    *error= true;
+  }
+  else
+  {
+    /* TODO: A non-alphanumeric alias name is invalid. */
+
+    /*
+      We have a valid alias name. Lets check if there is a value being
+      assigned to it. (Note: there mustn't be spaces around '='.)
+    */
+    name_len= strlen(name);
+
+    if (*pos == '=')
+    {
+      pos ++;
+      pos= parse_alias_value(pos, &value, &is_valid);
+
+      if (!is_valid)
+      {
+        tee_fprintf(stdout, "alias: '%s': invalid alias name\n", name);
+      }
+      else
+      {
+        /*
+          We now have a valid name and value, let add/update the
+          aliases hash.
+        */
+
+        if ((record= my_hash_search(&aliases, (const uchar *) name, name_len)))
+        {
+          my_hash_delete(&aliases, record);
+        }
+        alias= (ALIAS *) my_malloc(sizeof(ALIAS), MYF(MY_WME));
+        alias->name= name;
+        alias->name_len= name_len;
+        alias->value= value;
+        my_hash_insert(&aliases, (uchar *) alias);
+      }
+    }
+    else
+    {
+      /* Only alias name specified, lookup, print its value and return. */
+      alias= (ALIAS *) my_hash_search(&aliases, (const uchar*) name,
+                                      name_len);
+      if (alias)
+      {
+        tee_fprintf(stdout, "alias  %s = '%s'\n", alias->name, alias->value);
+      }
+      else
+      {
+        tee_fprintf(stdout, "alias: '%s': not found\n", name);
+      }
+    }
+    *error= false;
+  }
+  return pos;
+}
+
+static char *handle_next_unalias(char *line, bool *error)
+{
+  char *name, *pos;
+  uchar *record;
+  bool is_valid;
+
+  /* Parse the alias name. */
+  pos= parse_alias_name(line, &name, &is_valid);
+
+  if (!is_valid)
+  {
+    tee_fprintf(stdout, "unalias: '%s': invalid alias name\n", name);
+    *error= true;
+  }
+  else
+  {
+    if ((record= my_hash_search(&aliases, (const uchar *) name, strlen(name))))
+    {
+      my_hash_delete(&aliases, record);
+    }
+    else
+    {
+      tee_fprintf(stdout, "unalias: '%s': not found\n", name);
+    }
+    *error= false;
+  }
+  return pos;
```

Comment:
In current implementation, I guess you need to ```free(name)```. It'd be nice if ```parse_alias_name``` wouldn't do string copy.